### PR TITLE
Fix 2FA challenge and password challenge for non-database users

### LINKED ISSUE #11685: 2FA Challenge is not provided with LDAP authentication
### Expected behaviour

After login 2FA question should be provided

### Actual behaviour

Login without any 2FA is successful

### Steps to reproduce the problem

Setup LDAP auth in `.env.production`, login, enable OTP 2FA, logout
Than login again -> 2FA question not provided

### Specifications

Mastodon Version: v2.9.3

### Comments

There are no exceptions or warnings in the log.

### LINKED ISSUE #11691: Account deletion not working when logged in via LDAP account
### Expected behaviour

User which logged in via LDAP should be able to delete its Mastodon account (and only the Mastodon account, not the LDAP entry).

### Actual behaviour

User gets an error `Nice try, hackers! Incorrect password`

### Steps to reproduce the problem

Setup LDAP auth
Login as new LDAP user
Settings -> Account -> Delete account -> Enter password

### Specifications

Mastodon Version: v2.9.3

### PR DESCRIPTION
**Pull request by Gargron**
Fix #11685

Previous code (inspired by GitLab) handled 2FA *before* Devise performed its own authentication, duplicating the work, and, in case of non-database users (LDAP, PAM), doing the *wrong* kind of work. Now authentication is delegated to Devise/Warden, then once we have the user, we check the 2FA requirement, and handle OTP/recovery token.

Fix #11691

It is not possible to re-authenticate non-database users just from password input when confirming account deletion. So, for those users, use a simple username repetition challenge instead of confirming password.

### CODE REVIEW

**Comment by Gargron** [comment]
How did this break PAM authentication...

File: config/initializers/devise.rb [hunk 1]
```diff
@@ -73,8 +73,11 @@ def valid?
   config.warden do |manager|
     manager.default_strategies(scope: :user).unshift :ldap_authenticatable if Devise.ldap_authentication
     manager.default_strategies(scope: :user).unshift :pam_authenticatable  if Devise.pam_authentication
-    manager.default_strategies(scope: :user).unshift :two_factor_authenticatable
-    manager.default_strategies(scope: :user).unshift :two_factor_backupable
+    manager.default_strategies(scope: :user).unshift :database_authenticatable
```
Comment (Gargron) [inline config/initializers/devise.rb]: Ah, if I `unshift` it *after* the other stuff, it gets the first position, which is wrong. I see.

File: app/controllers/auth/sessions_controller.rb [hunk 2]
```diff
@@ -22,9 +20,22 @@ def new
...
 
   def create
-    super do |resource|
-      remember_me(resource)
-      flash.delete(:notice)
+    self.resource = begin
+      if user_params[:email].blank? && session[:otp_user_id].present?
+        User.find(session[:otp_user_id])
+      else
+        warden.authenticate!(auth_options)
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/controllers/auth/sessions_controller.rb]: It's fine to set resource... The sign in happens with the `sign_in` method.

[the remark this answers is not in the document]
Comment (ThibG) [reply app/controllers/auth/sessions_controller.rb]: Ok that seems fine.

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**